AP_Arming/AP_Scripting: allow scripts for pre-arm checks

rmackay9 opened the pull request:
This PR makes two somewhat unrelated changes:

- Resolves an issue with the onboard LOG message's "ArmChecks" field not displaying all the bits for the enabled checks.  No big deal of course but just so we know, it was this PR that introduced the issue: https://github.com/ArduPilot/ardupilot/pull/12810
- Add the ability for Scripts to stop the vehicle from arming

Below is a screen shot from a bench test where the example script was run on a CubeBlack.  This example script simply disallows arming the vehicle within 60 seconds of boot

![scripting-pre-arm](https://user-images.githubusercontent.com/1498098/74319116-c22d5880-4dc1-11ea-82e7-a624ae9ec55e.png)

Some known issues and thoughts on this change:

- In a follow-up PR I plan to extend support for companion computers to prevent arming using the [MAV_CMD_ARM_AUTHORIZATION_REQUEST](https://mavlink.io/en/messages/common.html#MAV_CMD_ARM_AUTHORIZATION_REQUEST) command
- Only 3 external "authorisers" are supported.  This low number is pretty arbitrary and could be increased easily with only minimal impact on RAM usage.  Each additional authoriser consumes about 42 bytes of RAM most of which is for the string to hold failure messages.
- We could modify the logic so that if there are any failed calls to get_external_auth_id() we disallow arming.  This would protect against cases where there are more than 3 scripts attempting to restrict arming
- There is a slight loophole in that we only protect against arming after the script has made a call to get_external_auth_id().  Theoretically it might be possible for a vehicle to become armed before the script has made the call to get_external_auth_id() although I think in practice this is not possible.

rmackay9 commented:
By the way, here is a screen shot of the ARM message before and after the fix.  For this test I set LOG_DISARM = 1 and ARMING_CHECK = 65536 (for camera).  We can see in the "before" that the ArmChecks field appears as "0" while in "after" it correctly shows "65536"
![ARM-log-fix](https://user-images.githubusercontent.com/1498098/74324639-aa0e0700-4dca-11ea-861e-25ac05167fd6.png)

rmackay9 commented:
I've found and fixed a couple of issues that I had missed.  I also added another example script but perhaps we don't need two example scripts and can drop one of the two.

rmackay9 commented on libraries/AP_Arming/AP_Arming.cpp:
```diff
@@ -850,6 +850,69 @@ bool AP_Arming::camera_checks(bool display_failure)
+    if (ext_auth_count < ext_auth_count_max) {
+        // allocate buffer for failure message
+        ext_auth[ext_auth_count].fail_msg = (char *)calloc(ext_auth_str_len, sizeof(char));
+        if (ext_auth[ext_auth_count].fail_msg == nullptr) {
```
yes, sounds good.  Added a new aux_auth_error member to track failures.

rmackay9 commented on libraries/AP_Arming/AP_Arming.h:
```diff
@@ -166,6 +173,22 @@ class AP_Arming {
+        ExtAuthStates state;
+        char* fail_msg;
+    };
+    ext_auth_t ext_auth[ext_auth_count_max] = {};
```
I don't think using a bitmask is worth the effort or additional complexity it would add.  The structure is only 7 bytes per authoriser so there's very limited room for shrinking the basic structure.
I'll switch it to a single failure message buffer but we should note that this comes at the cost of not being able to display all arming errors at once from independent scripts or companion computers.  I think it's very possible we could get an enhancement request to show all the failures at once

WickedShell replied:
6 bytes total would support 32 arming checks with my proposal though... Using the proposed structure here would be 224 bytes to support that many checks. 3 checks is just not enough for arming checks as far as I can tell, I'd easily like to have more then that many for scripts.

rmackay9 replied:
No, I think if we switch it to a single failure message and make the AuxAuthStates class based on a uint8_t it will consume just 32 bytes for 32 authorisers..

rmackay9 replied:
While modifying this to use a single fail_msg buffer, one issue is this situation:

1. authoriser 1 fails, set failure msg to "fail 1"
2. authoriser 2 fails, sets failure msg to "fail 2"
3. authoriser 2 passes

In this situation, authoriser 1 has failed but we've lost it's failure message but we need to tell the user that we've failed arming.. but can no longer say why

rmackay9 commented on libraries/AP_Arming/AP_Arming.cpp:
```diff
@@ -850,6 +850,69 @@ bool AP_Arming::camera_checks(bool display_failure)
+        ext_auth_count++;
+        return true;
+    }
+    return false;
```
txs, new commits treats this as an error case and arming will fail

rmackay9 commented on libraries/AP_Arming/AP_Arming.h:
```diff
@@ -81,6 +82,10 @@ class AP_Arming {
 
+    // external authorisation methods
+    bool get_external_auth_id(uint8_t& auth_id);
+    void set_external_auth_state(uint8_t auth_id, bool passed, const char* fail_msg);
```
I've split set_external_auth_state() into two methods, one for passed and one for fail.

rmackay9 commented on libraries/AP_Scripting/examples/arming-check.lua:
```diff
@@ -0,0 +1,20 @@
+
+function update() -- this is the loop which periodically runs
+  if auth_id then
+    now = millis()
```
sure, removed the timer based arming check script

rmackay9 commented on libraries/AP_Arming/AP_Arming.h:
```diff
@@ -166,6 +173,22 @@ class AP_Arming {
+
+    // external authorisation
+    static const uint8_t ext_auth_count_max = 3;
+    static const uint8_t ext_auth_str_len = 40;
```
sure, extended to be 4**2** bytes

rmackay9 commented on libraries/AP_Scripting/examples/arming-check-batt-temp .lua:
```diff
@@ -0,0 +1,19 @@
+  if auth_id then
+    now = millis()
+    batt_temp = battery:get_temperature(0)
+    if (batt_temp >= batt_temp_max) then
```
great, thanks.  fixed

rmackay9 commented on libraries/AP_Arming/AP_Arming.cpp:
```diff
@@ -850,6 +850,69 @@ bool AP_Arming::camera_checks(bool display_failure)
+        }
+    }
+    if (!some_failed && waiting_for_responses) {
+        check_failed(ARMING_CHECK_EXTAUTH, display_failure, "Waiting for external authorisation");
```
sure, renamed "external" to "auxiliary"

rmackay9 commented on libraries/AP_Arming/AP_Arming.cpp:
```diff
@@ -850,6 +850,69 @@ bool AP_Arming::camera_checks(bool display_failure)
 
+// request an external authorisation id.  This id should be used in subsequent calls to set_external_auth_state
+// returns true on success
+bool AP_Arming::get_external_auth_id(uint8_t& auth_id)
```
great, thanks. semaphore added

rmackay9 commented on libraries/AP_Arming/AP_Arming.cpp:
```diff
@@ -850,6 +850,69 @@ bool AP_Arming::camera_checks(bool display_failure)
+            if (fail_msg == nullptr) {
+                hal.util->snprintf(ext_auth[auth_id].fail_msg, ext_auth_str_len, "External authorisation refused");
+            } else {
+                hal.util->snprintf(ext_auth[auth_id].fail_msg, ext_auth_str_len, fail_msg);
```
txs, changed to strncpy

rmackay9 commented on libraries/AP_Arming/AP_Arming.cpp:
```diff
@@ -850,6 +850,69 @@ bool AP_Arming::camera_checks(bool display_failure)
+        } else {
+            ext_auth[auth_id].state = ExtAuthStates::AUTH_STATE_FAILED;
+            if (fail_msg == nullptr) {
+                hal.util->snprintf(ext_auth[auth_id].fail_msg, ext_auth_str_len, "External authorisation refused");
```
txs, again, changed to strncpy

rmackay9 commented on libraries/AP_Arming/AP_Arming.cpp:
```diff
@@ -850,6 +850,69 @@ bool AP_Arming::camera_checks(bool display_failure)
+// set external authorisation state
+void AP_Arming::set_external_auth_state(uint8_t auth_id, bool passed, const char* fail_msg)
+{
+    if ((auth_id < ext_auth_count_max) && (ext_auth[auth_id].state != ExtAuthStates::AUTH_STATE_UNASSIGNED)) {
```
txs, semaphore added

rmackay9 commented on libraries/AP_Arming/AP_Arming.cpp:
```diff
@@ -850,6 +850,69 @@ bool AP_Arming::camera_checks(bool display_failure)
+
+bool AP_Arming::external_auth_checks(bool display_failure)
+{
+    // check results for each external authorisation id
```
semaphore added..

rmackay9 commented:
I think I've addressed all of @WickedShell's comments now (thanks for the review!) and I think it's in better shape now.  Any further comments are very welcome!

WickedShell commented on libraries/AP_Arming/AP_Arming.h:
```diff
@@ -166,6 +174,21 @@ class AP_Arming {
+    static const uint8_t aux_auth_count_max = 3;    // maximum number of auxiliary authorisers
+    static const uint8_t aux_auth_str_len = 42;     // maximum length of failure message (50-8 for "PreArm: ")
+    enum class AuxAuthStates : uint8_t {
+        AUX_AUTH_STATE_UNASSIGNED = 0,
```
This state doesn't matter, you assign them in a strictly increasing order based on `auth_count` which means any ID less then `auth_count` *must* be in the NO_RESPONSE, FAILED, or PASSED state.

WickedShell replied:
Also with an enum class like this we shouldn't be prefixing `AUX_AUTH_STATE` to all the members. It's made all the users incredibly verbose and less clear whats happening.

rmackay9 replied:
ok, removed the no-response and shortened the names of the states to NO_RESPONSE, AUTH_FAILED and AUTH_PASSED

WickedShell commented on libraries/AP_Arming/AP_Arming.cpp:
```diff
@@ -850,6 +850,116 @@ bool AP_Arming::camera_checks(bool display_failure)
+        }
+    }
+    auth_id = aux_auth_count;
+    aux_auth_state[auth_id] = AuxAuthStates::AUX_AUTH_STATE_NO_RESPONSE;
```
This is unneeded once this state is removed.

rmackay9 replied:
ok, removed

WickedShell commented on libraries/AP_Arming/AP_Arming.cpp:
```diff
@@ -850,6 +850,116 @@ bool AP_Arming::camera_checks(bool display_failure)
+    // update state
+    aux_auth_state[auth_id] = AuxAuthStates::AUX_AUTH_STATE_FAILED;
+
+    // store failure message if this authoriser has the lowest auth_id
```
Why bother with storing the lowest ID first? The order in which scripts get loaded here is not well defined, so there is nothing special about ID 0 before ID 1. Why not just always store the most recent failure one here? That has the merits of being both less code and stores the failure that we know the most about. If you had a script that was checking GPS information at 0.2Hz and a different script monitoring the status of some RC channel at 1Hz as an example, the GPS one may be true now and just waiting for the script to run, but if the RC channel is asserting its failed the most recently then this is the most likely thing for a user to need to fix.

rmackay9 replied:
I think I'll leave this one as it is which leads to the first authoriser's message being displayed until it passes.  If we don't do this then we end up with messages from various authorisers being somewhat randomly reported.  I.e. it could easily display the failure from the first authoriser, then a moment later display the 2nd authoriser's failure message, then back to the first.

I think we should either treat all the authorisers as if they're part of the same subsystem (i.e. display the first message until it passes, then the 2nd, etc) or we treat them as if they're all different subsystems in which case we go back to the original plan of having separate message buffers for each authoriser.

WickedShell commented on libraries/AP_Arming/AP_Arming.cpp:
```diff
@@ -850,6 +850,116 @@ bool AP_Arming::camera_checks(bool display_failure)
+        case AuxAuthStates::AUX_AUTH_STATE_PASSED:
+            break;
+        }
+    }
```
You need to do a `check_failed` if there is an ID which is failing, but we are missing the string for it. At the moment its possible the check we had that set the failure string could have been set to passing, but another check is still failing, but hasn't updated itself recently, which means the failure string is stale.

rmackay9 replied:
So the way it's written now, it will simply not display the message if it's stale but this is probably not good behaviour because it's been told to "display_failure" so I'll make it send a generic failure message like, "Waiting for auxiliary authorisation".

I'm not keen on fixing it further because, as discussed on the dev call last week, I suspect we will end up going back to adding authoriser independent failure messages which washes away all these issues.

WickedShell commented on libraries/AP_Scripting/generator/description/bindings.desc:
```diff
@@ -41,6 +41,9 @@ singleton AP_Arming alias arming
 singleton AP_Arming method is_armed boolean
 singleton AP_Arming method arm boolean AP_Arming::Method::SCRIPTING'literal
+singleton AP_Arming method get_aux_auth_id boolean uint8_t'Null
+singleton AP_Arming method set_aux_auth_passed void uint8_t 0 255
```
Why not expose `aux_auth_count_max` as a public member of AP_Arming so that you can use it here for bounds checking from a script? I know it's bound checked inside of AP_Arming, but the advantage of the check here (which we will be running either way) is that if its a larger ID it actually throws a meaningful error in the script that the argument was out of range, which lets people check what happened, rather then thinking it's just not working.

Also `UINT8_MAX` rather then 255 :)

rmackay9 replied:
With the removal of the AUX_AUTH_STATE_UNASSIGNED, I've changed the range checking within AP_Arming to use aux_auth_count (the number of allocated authorisers).

Does this kind of dynamic range checking work?  I.e. I create a function that returns a uint8 and it will be used before each call to set_aux_auth_passed?

I thought we were trying to keep the number of bindings small because each one consumes a lot of memory..

I've changed the limits to use UINT8_MAX, thanks.

rmackay9 replied:
I tried but failed to get the "ud->get_aux_auth_count()" to work for range checking the input to the other functions (it always failed) so I'll leave this suggestion out.  I think we can probably all agree it's nice to have but not a necessity.

rmackay9 commented:
I've rebased on master and re-tested and it seems to be working still.  I'll merge once it passes travis.

Thanks again for the review

WickedShell commented:
@rmackay9 If you just merge a PR after making changes that defeats all the checks the review process is supposed to be bringing you. The point of the previous review is there were things which has to be improved. I'm happy you addressed them, but after making changes a review should still be done again. Just instant merging after you've pushed up changes to address a review defeats the quality of the review process.

Particularly since you didn't understand one of the review comments, and just merged over it anyways, without waiting for any discussion on how to properly address it. This is extremely frustrating as someone doing reviews if we can't achieve a consensus while it's in the PR stage then why bother presenting reviews and try to improve the quality of new features.

rmackay9 commented:
@WickedShell, Yes, I understand your point of view but please also understand mine:

- The PR has been open for review for two weeks and I've quickly addressed all issues raised
- I've been very careful about testing it thoroughly and providing proof of the testing
- It's already been approved by one of the other senior devs

I very much appreciate the review you provided and it did catch at least one real issue but my feeling is that we were beyond catching any real bugs and were at the point of differences in style.

As a side note, I think scripting is a really important feature and we should try to speed up those merges that are simply adding bindings which should be pretty safe.
